File: cs50x/readability/readability.c
// Program to grade level a text.
// Could be More efficient by combining the functions into one function that would scan the text only--
// -- once and return all values. Cs50 implementaion requires there to be 3 functions ,however, so I did so.

#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

//All functions take the text and its length
int noletters(string t, int l);
int nowords(string t, int l);
int nosents(string t, int l);


int main(void)
{
    // Get string of text
    string text = get_string("Write text:\n");

    // Get length of string
    int length = strlen(text);

    // calculate number of sentences (punct marks
    //uses number of words to add the words at the end of a sentence
    int sents = nosents(text, length);

    // calculate number of words (number of blank spaces + 1 because of last word in last sentence won't have a space after it )
    int words = nowords(text, length);

    // Calculate number of letters
    int letters = noletters(text, length);

    // Sugar of Programing
    if (letters == 0 || sents == 0 || words == 0)
    {
        printf("ERORR: Please make sure your write at least one word, and use punctuation marks\n");
        return 1;
    }

    // Calculate average of letters per 100 words in text (L)
    float L = (float)letters / words * 100;

    // Calculate average of sentences per 100 words.
    float S = (float)sents / words * 100;

    // Apply and print the index number rounded
    int index = round(0.0588 * L - 0.296 * S - 15.8);
    if (index < 16 && index >= 1)
    {
        printf("Grade %i\n", index);
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Something went wrong!\n");
        return 1;
    }

}

//FUNCTIONS:------
//Calculates number of letters in a text
int noletters(string t, int l)
{
    int letters = 0;
    for (int i = 0; i < l; i++)
    {
        if (isalpha(t[i]))
        {
            letters++;
        }
    }
    return letters;
}

// Calculates the number of words in a text
int nowords(string t, int l)
{
    int words = 0;
    for (int i = 0; i < l; i++)
    {
        if (isspace(t[i]))
        {
            words++;
        }
    }
    // Last word in text (because it has only a punct mark after it and no space) wont be recorded, so we add 1
    return words + 1;
}

//Calculates number of sentences in a text.
int nosents(string t, int l)
{
    int sents = 0;
    for (int i = 0; i < l; i++)
    {
        if (t[i] == '.' || t[i] == '?' || t[i] == '!')
        {
            sents++;
        }
    }
    return sents;
}

File: cs50x/sentimental-readability/readability.py
import sys


# Dict to store values
main_dict = {"letter_count": 0, "sent_count": 0, "word_count": 0}


def main():

    # Get text and length
    text = input("Input Text:\n")
    length = len(text)

    # Calculate letters, words, and sentences in text. (Stored in main_dict)
    num_variables(text, length)
    sent_count = main_dict["sent_count"]
    word_count = main_dict["word_count"]
    letter_count = main_dict["letter_count"]

    # Catching errors
    if (letter_count == 0 or sent_count == 0):
        print("ERORR: Please make sure your write at least one word, and use punctuation marks")
        sys.exit(1)

    # Average of letters per 100 words
    letter_avg = (letter_count / word_count) * 100

    # Average of sentences per 100 words
    sent_avg = (sent_count / word_count) * 100

    # Apply and print the index number rounded
    index = round(0.0588 * letter_avg - 0.296 * sent_avg - 15.8)
    if (index < 16 and index >= 1):
        print(f"Grade {index}")

    elif (index >= 16):
        print("Grade 16+")

    elif (index < 1):
        print("Before Grade 1")

    else:
        print("Something went wrong!")
        sys.exit(2)


def num_variables(text, length):
    # Temp variables for counting. words starts at 1 because the last word, the one that doesn't have a space after the end of the sentence, wont be recorded otherwise.
    l = 0
    w = 1
    s = 0
    for i in range(length):
        if (text[i].isalpha()):
            l += 1

        elif (text[i].isspace()):
            w += 1

        elif (text[i] == '.' or text[i] == '?' or text[i] == '!'):
            s += 1

    main_dict["letter_count"] = l
    main_dict["word_count"] = w
    main_dict["sent_count"] = s


main()